Use TruffleLogger bound to the engine to prevent illegal usage

hubertp opened the pull request:
### Pull Request Description

Using a `TruffleLogger` in `SerializationManager` that is bound to the engine rather than the context prevents reaching an illegal state when using thread pools. 

Also cleaned up some tests for consistency. 

To verify the fix
```diff
--- a/engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala
+++ b/engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala
@@ -31,7 +31,7 @@ final class SerializationManager(compiler: Compiler) {
   import SerializationManager._
 
   /** The debug logging level. */
-  private val debugLogLevel = Level.FINE
+  private val debugLogLevel = Level.INFO
```
and run
`sbt:enso> runtime/test`

Closes #8147.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.

hubertp committed:
Provide log level to prevent truffle logger crash

Partial fix for Truffle logger crashes - context required log levels.
With a suggested change of log level in `SerializationManager`
`sbt:enso> runtime/testOnly *SerdeCompilerTest`
now passes but
`sbt:enso> runtime/test`
...will continue to fail in `SerdeCompilerTest`, which is really
confusing.

JaroslavTulach commented on engine/runtime/src/test/java/org/enso/compiler/SerdeCompilerTest.java:
```diff
@@ -115,6 +116,7 @@ private Context ensoContextForPackage(String name, File pkgFile, boolean disable
             .option(
                 RuntimeOptions.LANGUAGE_HOME_OVERRIDE,
                 Paths.get("../../distribution/component").toFile().getAbsolutePath())
+            .option(RuntimeOptions.LOG_LEVEL, Level.WARNING.getName())
```
I don't understand what is the idea of the fix? I don't see any change in the code, just in the tests. Thus the only idea I see is: _Increase logging level, so logging doesn't happen_. But I wanted the logging to happen when I faced the issue.

hubertp replied:
Yes, this was only a _partial_ fix for some of the tests. It wasn't addressing the core of the problem. Now it does.

hubertp committed:
Use TruffleLogger bound to the engine

TruffleLoggers bound to the context were problematic when being invoked
in threads coming from a thread pool. By explicitly using a
TruffleLogger that is bound to the engine we can prevent such illegal
usage.

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java:
```diff
@@ -40,14 +40,14 @@
     this.context = context;
-    this.compiler = context.getLogger(Compiler.class);
-    this.serializationManager = context.getLogger(SerializationManager.class);
+    this.compilerLogger = context.getLogger(Compiler.class);
```
I left this logger as-is as I haven't seen the same illegal state failures as with SerializationManager, potentially because (de-)serializationis is typically run in a separate thread. We could change it as well, I'm open for suggestions.

JaroslavTulach replied:
I think we want to bound all loggers to engine. No need to have two `getLogger` and `getLoggerBoundToEngine` method.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoContext.java:
```diff
@@ -774,6 +774,17 @@ public TruffleLogger getLogger(Class<?> klass) {
+   * @return a new logger for the specified {@code path}
+   */
+  public TruffleLogger getLoggerBoundToEngine(Class<?> clazz) {
+    return environment.getLogger(clazz);
```
Let's just use this in the existing `getLogger` method without introducing a new one.

JaroslavTulach replied:
I guess this is the core of the fix. It seems correct.

JaroslavTulach commented on engine/runtime/src/test/java/org/enso/compiler/ExecCompilerTest.java:
```diff
@@ -28,6 +29,10 @@ public static void initEnsoContext() {
         )
         .option(RuntimeOptions.STRICT_ERRORS, "false")
+        .option(
+                RuntimeOptions.LOG_LEVEL,
```
Why are these `LOG_LEVEL` changes needed?

hubertp replied:
Otherwise we rely on the parent logger to get the level, which is not obvious what it is. I also made sure that we no longer ignore the output in various places.

JaroslavTulach replied:
Default configuration of `java.util.logging` is to log `Level.INFO` or higher to console. Not sure whether that applies to `TruffleLogger`, but I assume so, it was created with intention to mimic `java.util.logging`.

hubertp replied:
That could be right, I wasn't able to verify it. I see no harm in specifying log level explicitly, other than being a bit more verbose. But if that's a dealbreaker for you, I can remove it.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoContext.java:
```diff
@@ -774,6 +774,17 @@ public TruffleLogger getLogger(Class<?> klass) {
+   * @param clazz the class to name log entries with
+   * @return a new logger for the specified {@code path}
+   */
+  public TruffleLogger getLoggerBoundToEngine(Class<?> clazz) {
```
Unrelated, but: Why not return `slf4j` logger from this method?

hubertp replied:
I tried **not** to use `slf4j` within runtime at all. To not tempt people to use it.

JaroslavTulach replied:
Using it as a simple facade over `TruffleLogger` would at least give us the same formatting of messages.

JaroslavTulach reviewed (commented):
I don't see a reason to have `getLogger` and `getLoggerBoundToEngine` next to each other - especially when there is 1:1 mapping between `Engine` and `EnsoContext`.

hubertp committed:
Don't ignore relevant logs

Logs were often sent to null outpu stream or `ByteArrayOutputStream`,
ending up in a black hole. We want to see the logs above the specific
level because they are relevant to debugging problems.

JaroslavTulach commented on engine/runtime-with-instruments/src/test/java/org/enso/interpreter/test/instrument/AvoidIdInstrumentationTagTest.java:
```diff
@@ -38,7 +40,11 @@ public void initContext() {
+                RuntimeOptions.LOG_LEVEL,
+                Level.WARNING.getName()
+        )
+        .logHandler(System.err)
```
Doesn't this pollute the `test` command output with a lot of debug messages? Ideally they should only be displayed when the test fails. There is no point in showing logged messages when the test succeeds, I think.

hubertp replied:
Nope. I have not seen a lot of, if any, warnings being reported in our test suite.

hubertp committed:
nit

hubertp committed:
nit

JaroslavTulach reviewed (approved):
Using `environment.getLogger(clazz)` seems like correct fix.
